src/index/FTSAlgorithms.cpp

```diff
@@ -1264,18 +1185,28 @@ void FTSAlgorithms::multVarsFilterAggScoresAndTakeTopKContexts(
   vector<Id> filteredEntitiesInContext;
   TextRecordIndex currentCid = wep.cids_[0];
   Score cscore = wep.scores_[0];
-  vector<WordIndex> cwids;
-  cwids.resize(numOfTerms);
-  for (size_t l = 0; l < numOfTerms; l++) {
-    cwids[l] = wep.wids_[l].empty() ? 0 : wep.wids_[l][0];
+  ad_utility::HashMap<Id, std::set<vector<WordIndex>>> currentCombinedWids;
+  vector<WordIndex> wids;
+  wids.resize(numTerms);
+  for (size_t l = 0; l < numTerms; l++) {
+    AD_CONTRACT_CHECK(wep.wids_[l].size() == wep.cids_.size());
+    wids[l] = wep.wids_[l][0];
   }
+  currentCombinedWids[wep.eids_[0]].insert(wids);
 
   for (size_t i = 0; i < wep.cids_.size(); ++i) {
     if (wep.cids_[i] == currentCid) {
-      if (fMap.contains(wep.eids_[i])) {
-        filteredEntitiesInContext.push_back(wep.eids_[i]);
+      for (size_t l = 0; l < numTerms; l++) {
+        wids[l] = wep.wids_[l][i];
       }
-      entitiesInContext.push_back(wep.eids_[i]);
+      if (entitiesInContext.empty() ||
+          entitiesInContext.back() != wep.eids_[i]) {
+        if (fMap.contains(wep.eids_[i])) {
+          filteredEntitiesInContext.push_back(wep.eids_[i]);
+        }
+        entitiesInContext.push_back(wep.eids_[i]);
+      }
+      currentCombinedWids[wep.eids_[i]].insert(wids);
```

Comment:
We can discuss this together, You can quickly explain to me what is going on here.